### PR TITLE
bpo-42264: Deprecate sqlite3.OptimizedUnicode

### LINKED ISSUE #86430: Deprecate or remove sqlite3.OptimizedUnicode
BPO | [42264](https://bugs.python.org/issue42264)
--- | :---
Nosy | @vstinner, @berkerpeksag, @serhiy-storchaka, @pablogsal, @erlend-aasland
PRs | <li>python/cpython#23163</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2020-11-17.16:20:23.996>
created_at = <Date 2020-11-04.21:24:26.601>
labels = ['library', '3.10']
title = 'Deprecate or remove sqlite3.OptimizedUnicode'
updated_at = <Date 2020-11-17.16:20:23.995>
user = 'https://github.com/erlend-aasland'
```

bugs.python.org fields:
```python
activity = <Date 2020-11-17.16:20:23.995>
actor = 'serhiy.storchaka'
assignee = 'none'
closed = True
closed_date = <Date 2020-11-17.16:20:23.996>
closer = 'serhiy.storchaka'
components = ['Library (Lib)']
creation = <Date 2020-11-04.21:24:26.601>
creator = 'erlendaasland'
dependencies = []
files = []
hgrepos = []
issue_num = 42264
keywords = ['patch']
message_count = 2.0
messages = ['380363', '381252']
nosy_count = 5.0
nosy_names = ['vstinner', 'berker.peksag', 'serhiy.storchaka', 'pablogsal', 'erlendaasland']
pr_nums = ['23163']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue42264'
versions = ['Python 3.10']
```

</p></details>

### PR DESCRIPTION
**Pull request by erlend-aasland**
Fixes #86430

### CODE REVIEW

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/sqlite3/deprecated.py
Comment (serhiy-storchaka) [inline Lib/sqlite3/deprecated.py]: Add `stacklevel=2`.

You could also check the file name of the emitted warning in tests.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/sqlite3/deprecated.py
Comment (erlend-aasland) [reply Lib/sqlite3/deprecated.py]: Of course, thanks! Fixed in a9e959f

**Comment by erlend-aasland** [comment]
I have made the requested changes; please review again

File: Lib/sqlite3/deprecated.py
Comment (erlend-aasland) [reply Lib/sqlite3/deprecated.py]: Could've used `os.path.basename(__file__)` iso. `factory.py` in the unit test, but this test is bound to be short-lived anyway.

Comment (serhiy-storchaka) [reply Lib/sqlite3/deprecated.py]: Maybe just compare filename with `__file__`?

In any case, as you said, this is a short-lived test, so it does not matter.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by erlend-aasland** [comment]
Thanks for reviewing, @serhiy-storchaka!

**Comment by serhiy-storchaka** [comment]
And it could be good idea to add tests for filename to all existing deprecation tests.

**Comment by erlend-aasland** [comment]
> And it could be good idea to add tests for filename to all existing deprecation tests.

I agree. I can open an issue for it.

**Comment by erlend-aasland** [comment]
> And it could be good idea to add tests for filename to all existing deprecation tests.

There's a lot:
```
bash$ grep -r "DeprecationWarning" Lib/test | grep -v " as " | wc -l                                                                                                               (master)cpython.git
     519
```

File: Lib/sqlite3/deprecated.py
Comment (vstinner) [inline Lib/sqlite3/deprecated.py]: Can't you put the code directly in sqlite3/__init__.py?

File: Misc/NEWS.d/next/Library/2020-11-05-13-32-41.bpo-42264.r4KYUU.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2020-11-05-13-32-41.bpo-42264.r4KYUU.rst]:
```suggestion
``sqlite3.OptimizedUnicode`` has been deprecated. It has been undocumented
and obsolete since Python 3.3.
```

File: Doc/whatsnew/3.10.rst
Comment (vstinner) [inline Doc/whatsnew/3.10.rst]: Can you announce when it will be removed? https://www.python.org/dev/peps/pep-0387/ prevents to remove it before Python 3.12. So I suggest Python 3.12.

Comment (vstinner) [reply Doc/whatsnew/3.10.rst]: You may also explain that it is simply an alias to the builtin str type.

File: Lib/sqlite3/deprecated.py
Comment (erlend-aasland) [reply Lib/sqlite3/deprecated.py]: Sure. I used a separate file to make it more explicit.

File: Doc/whatsnew/3.10.rst
Comment (erlend-aasland) [reply Doc/whatsnew/3.10.rst]: Of course! (I was not aware of PEP 387; thanks!)

Comment (erlend-aasland) [reply Doc/whatsnew/3.10.rst]: How about this? b106c9e

**Comment by erlend-aasland** [comment]
PTAL, @vstinner

File: Lib/sqlite3/__init__.py
Comment (erlend-aasland) [inline Lib/sqlite3/__init__.py]:
```suggestion
            OptimizedUnicode is deprecated and will be removed in Python 3.12.
            Since Python 3.3 it has simply been an alias for 'str'.
```
Mind if I change the wording here as well, @vstinner?

Comment (vstinner) [reply Lib/sqlite3/__init__.py]: Do whatever you want, it's your PR :-)

Comment (erlend-aasland) [reply Lib/sqlite3/__init__.py]: All right, I'm done! No more finishing touches :)

**Comment by erlend-aasland** [comment]
FYI, Ubuntu CI hangs again:
`test_virtual_event (tkinter.test.test_ttk.test_widgets.ComboboxTest) ... Timeout (0:20:00)!`

**Comment by vstinner** [comment]
>  test_virtual_event (tkinter.test.test_ttk.test_widgets.ComboboxTest) ... Timeout (0:20:00)!

It is a know issue https://bugs.python.org/issue42142

You can ignore it. When it happens, I simply re-run the job until it pass.

File: Misc/NEWS.d/next/Library/2020-11-05-13-32-41.bpo-42264.r4KYUU.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2020-11-05-13-32-41.bpo-42264.r4KYUU.rst]: I suggest to copy your good What's New entry:

```

  ``sqlite3.OptimizedUnicode`` has been undocumented and obsolete since Python
  3.3, when it was made an alias to :class:`str`.  It is now deprecated,
  scheduled for removal in Python 3.12.
```

File: Lib/sqlite3/__init__.py
Comment (vstinner) [inline Lib/sqlite3/__init__.py]: Please add a reference to the bpo-42264:

```suggestion
# bpo-42264: OptimizedUnicode was deprecated in Python 3.10.  It's scheduled for removal
```

I let you reformat the comment if needed.

**Review by vstinner — COMMENTED** [review]
Good, it's way better! More comments.

File: Misc/NEWS.d/next/Library/2020-11-05-13-32-41.bpo-42264.r4KYUU.rst
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2020-11-05-13-32-41.bpo-42264.r4KYUU.rst]: Agreed! Fixed in 2507fd03301a47ee4df6eb45cbe99afbf322ab12

File: Lib/sqlite3/__init__.py
Comment (erlend-aasland) [reply Lib/sqlite3/__init__.py]: Good idea. Fixed in fe7c65e76213f9b4565ef3eba37008e60f2d834f

**Comment by erlend-aasland** [comment]
Please take another look, @vstinner.

**Review by vstinner — APPROVED** [review]
LGTM.

@serhiy-storchaka: Do you want to double check this PR?

**Comment by erlend-aasland** [comment]
@serhiy-storchaka Will you merge this after double checking, or will that be up to another core dev to decide?

**Review by serhiy-storchaka — APPROVED** [review]

**Review by hafizulsaat92 — APPROVED** [review]

**Pull request merged**
